### ISSUE TITLE
question about loss.py

### ISSUE THREAD

**Issue by jfzhang95** [1991viet/Yolo-pytorch#7]
Hello, thank you for your wonderful work. When I read the loss function, I cannot understand this line
https://github.com/1991viet/Yolo-pytorch/blob/b009df4b86815edd2d05f1f882674a2e6b6eaaf5/src/loss.py#L134
In my perspective, the size of iou_gt_pred is [num_gt_bbox, num_anchors]. The num_anchors is 5 x height x width. So I think `best_n` is between 0 to 5 x height x width - 1. Then will the `best_n * height * width + gj * width + gi` out of index?

Any feedback would be helpful. Thank you very much!

**Comment by 1991viet** [comment]
Hi,
Thanks for your interest on my code. Let me explain:
For each image, we build (5 * height * width) anchors, you are right.
In line 127, you could see, for each anchor, we find out **index** of the ground truth which it overlaps most (**best_anchors** here are actually index of best ground truth, if you have 2 persons, 2 cats and 2 dogs in an image, then you have 6 ground truths, and the index will range from 0 to 5), resulting in **iou_gt_pred** of size (6, 5 * height * width). As a result, (best_n * height * width + gj * width + gi) will always fall within [0, 5*height*width).
I hope you got it. If you have any further question, please tell me.

**Comment by jfzhang95** [comment]
Yes, I got it. Thank you for your explanation!

**Comment by jfzhang95** [comment]
Sorry, I am still confused. In the example you mentioned before, best_n ranges from 0 to 5. Then when best_n=5, `(best_n * height * width + gj * width + gi)` will out of index. Did I misunderstand something here? Hope for your help, thank you!

**Comment by 1991viet** [comment]
Hi,
We have 6 ground truths, then the best_n ranges from 0 to 5. If we have 10 ground truth, the best_n ranges from 0 to 9, and so on.  
For each image, we generate (5 * height * width) anchors, the shape of **iou_gt_pred** is (num_ground_truth, 5 * height * width) (num_ground_truth = 6 means n_best is 0-5, num_ground_truth = 10 means n_best is 0-9)
Bests,

**Comment by jfzhang95** [comment]
Yes, I know what you mean. But what confused me is when an anchor's best_n is 5 (suppose number of gt object is 6), then best_n(=5) * width * height + gj * width + gi will be larger than 5 * height * width. Maybe I misunderstand in somewhere. Thank you for your patient and help.

**Comment by jfzhang95** [comment]
Plus, since iou_gt_anchors = bbox_ious(gt_wh, anchors) and best_anchors = iou_gt_anchors.max(1)[1], best_anchors' size should be [num_gt_bbox] and each value of it ranges from 0 to 5 * height * width - 1.

I got this result through setting gt_wh = torch.rand(10, 4) and anchors = torch.rand(100, 4) and got best_anchor of size [10] and each value of it was lesser than 100.

Hope for your reply. Thank you very much!

**Comment by jfzhang95** [comment]
Hi, 
my problem is solved. It is my fault to misunderstand iou_gt_pred and iou_gt_anchors. 

Thank you again for your wonderful work!

**Comment by 1991viet** [comment]
You are welcome
